src/test/java/io/vertx/mqtt/test/client/MqttClientOutOfOrderAcksTest.java

```diff
@@ -0,0 +1,160 @@
+  }
+
+  @Test
+  public void publishQoS2OutOfOrderAcks(TestContext context) throws InterruptedException {
+    clientSendThreePublishMessages(MqttQoS.EXACTLY_ONCE, context);
+  }
+
+  private void clientSendThreePublishMessages(MqttQoS mqttQoS, TestContext context) {
+    Async async = context.async(3);
+    MqttClient client = MqttClient.create(vertx);
+
+    Queue<Integer> expectOrder = new LinkedList<>();
+    // order we expect to receive acknowledgment for published message
+    expectOrder.add(2);
+    expectOrder.add(1);
+    expectOrder.add(3);
+
+
+    client.publishCompletionHandler(h -> {
+      context.assertTrue(h.intValue() == expectOrder.poll());
+      log.info("[CLIENT] Publish completed for message with id: " + h);
+      async.countDown();
+    });
+
+    client.connect(MqttClientOptions.DEFAULT_PORT, MqttClientOptions.DEFAULT_HOST, c -> {
+
+      // publish QoS = 1 message three times
+      for (int i = 0; i < 3; i++)
+        client.publish(MQTT_TOPIC,
+          Buffer.buffer(MQTT_MESSAGE.getBytes()),
+          mqttQoS,
+          false,
+          false, h -> log.info("[CLIENT] publishing message id = " + h.result()));
+    });
+
+    async.await();
+    client.disconnect();
+  }
+
+
+  @Before
+  public void before() {
+    server = MqttServer.create(vertx);
+    server.endpointHandler(MqttClientOutOfOrderAcksTest::serverLogic).listen(ar -> {
+      if (ar.succeeded()) {
+        log.info("[SERVER] MQTT server listening on port " + ar.result().actualPort());
+      } else {
+        log.error("[SERVER] Error starting MQTT server", ar.cause());
+        System.exit(1);
+      }
+    });
+
+    server.exceptionHandler(t -> context.assertTrue(false));
+  }
+
+  @After
+  public void after() {
+    this.server.close();
+    this.vertx.close();
+  }
+
+  private static void serverLogic(MqttEndpoint mqttEndpoint) {
+    log.info("[SERVER] Client connected");
+
+    mqttEndpoint.publishHandler(p -> {
+      log.info("[SERVER] Received PUBLISH with message id = " + p.messageId());
+
+      if (p.qosLevel().equals(MqttQoS.EXACTLY_ONCE))
+        // when received last PUBLISH message to acknowledge
+        if (p.messageId() == 3) {
+          // send PUBREC in order 3 --> 2 --> 1 -->
+          mqttEndpoint.publishReceived(3);
+          mqttEndpoint.publishReceived(2);
+          mqttEndpoint.publishReceived(1);
+        }
+
+      if (p.qosLevel().equals(MqttQoS.AT_LEAST_ONCE))
+        // when received last PUBLISH message to acknowledge
+        if (p.messageId() == 3) {
+          //puback in order 2 --> 1 --> 3 -->
+
+          mqttEndpoint.publishAcknowledge(2);
+          mqttEndpoint.publishAcknowledge(1);
+          mqttEndpoint.publishAcknowledge(3);
+        }
+    });
+
+    mqttEndpoint.publishReleaseHandler(pr -> {
+      log.info("[SERVER] Receive PUBREL with message id = " + pr);
+
+      // when received last PUBREL message to acknowledge
+      if (pr == 1) {
```

Comment:
Related to the previous question it should be `pr == 3` if we change to 2,1 and 3 the PUBREC order ?